### PR TITLE
Implement QuicStreamBufferAllocator

### PR DESCRIPTION
**Pull request by danzh2010**
Implement this API with default SimpleBufferAllocator which use bare new and delete. In light of flag-enabled tcmalloc support, the default implementation will use tcmalloc underneath.

Roll up quiche tar ball to ba6354aa1b39f3d9788ead909ad3e678ac863938

Risk Level: low, not in use
Testing: Added simple test in test/extensions/quic_listeners/quiche/platform:quic_platform_test
Part of #2557

### CODE REVIEW

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -236,8 +236,8 @@ REPOSITORY_LOCATIONS = dict(
         urls = ["https://github.com/google/subpar/archive/1.3.0.tar.gz"],
     ),
     com_googlesource_quiche = dict(
-        # Static snapshot of https://quiche.googlesource.com/quiche/+archive/4fbea5de9afdf30611b27afd54c45a596944f9c2.tar.gz
-        sha256 = "2cf9f5ea62a03ca0d8773fe4f56949b72c28ac5b1bcf43d850a571f4e32add2a",
-        urls = ["https://storage.googleapis.com/quiche-envoy-integration/4fbea5de9afdf30611b27afd54c45a596944f9c2.tar.gz"],
+        # Static snapshot of https://quiche.googlesource.com/quiche/+archive/553a966443258b103c97326f94b53b03b933361d.tar.gz
```
Comment (wu-bin) [inline bazel/repository_locations.bzl]: Can you help me update it to a version at or after https://quiche.googlesource.com/quiche/+/a27fd44c5c53a67f3fa772b805363eff039411d5?

File: bazel/external/quiche.BUILD [hunk 2]
```diff
@@ -209,6 +210,14 @@ cc_library(
     deps = [":quic_platform"],
 )
 
+cc_library(
+    name = "quic_buffer_allocator_interface_lib",
```
Comment (wu-bin) [inline bazel/external/quiche.BUILD]: How about "quic_buffer_allocator_lib", which is the name used internally?

File: source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
```
Comment (wu-bin) [inline source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h]: #include <stdlib.h> for malloc&free.

File: bazel/external/quiche.BUILD (hunk 2 above)
Comment (jmarantz) [reply bazel/external/quiche.BUILD]: +1 -- most build targets in envoy end with "_lib" or "_interface", not both.

File: source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h (hunk 3 above)
Comment (jmarantz) [reply source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h]: or cstdlib, which is guaranteed idempotent.

File: source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_stream_buffer_allocator.h
+class QuicStreamBufferAllocatorImpl : public QuicBufferAllocator {
+public:
+  ~QuicStreamBufferAllocatorImpl() override {}
+
+  char* New(size_t size) override { return static_cast<char*>(malloc(size)); }
```
Comment (jmarantz) [inline source/extensions/quic_listeners/quiche/platform/quic_stream_buffer_allocator_impl.h]: above these put `// QuickBufferAllocator` per envoy convention.

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 5]
```diff
@@ -531,5 +532,13 @@ TEST_F(FileUtilsTest, ReadFileContents) {
   EXPECT_EQ(data, output);
 }
 
+TEST_F(QuicPlatformTest, TestEnvoyQuicBufferAllocator) {
+  QuicStreamBufferAllocator allocator;
+  char* p = allocator.New(1024);
+  EXPECT_NE(nullptr, p);
+  memset(p, 'a', 1024);
+  allocator.Delete(p);
```
Comment (jmarantz) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: You could also in this test ensure that the total allocated bytes was the same before and after this delete, via `Memory::Stats::totalCurrentlyAllocated()` before and after. See test/common/stats/thread_local_store_test.cc , test HeapStatsThreadLocalStoreTest.MemoryWithoutTls for an example. Only do that test if TestUtil::hasDeterministicMallocStats() is true.

**Comment by jmarantz** [comment]
@wu-bin that's the way tcmalloc works: it overrides malloc and free as well as operator new and delete.

It's worth noting that compiling with tcmalloc enabled is controlled by bazel flags, and I think it might not be enabled on some platform (Mac?)

**Comment by wu-bin** [comment]
> @wu-bin that's the way tcmalloc works: it overrides malloc and free as well as operator new and delete.
> 
@jmarantz That's true, but other memory allocators also overrides malloc/free etc.

> It's worth noting that compiling with tcmalloc enabled is controlled by bazel flags, and I think it might not be enabled on some platform (Mac?)

**Comment by jmarantz** [comment]
@wu-bin RE "other memory allocators" -- you're right of course :) The implementation here doesn't really depend on tcmalloc and doesn't claim it does, except in the PR description. It's just delegating to the system memory allocator, which seems fine to me. @danzh2010 maybe just change the description to reflect that?

**Comment by danzh2010** [comment]
> @wu-bin that's the way tcmalloc works: it overrides malloc and free as well as operator new and delete.
Quic has another allcator: QuicSimpleBufferAllocator which use bare ::new and ::delete. Could this two ends up with doing the same thing underneath?

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (wu-bin) [reply bazel/repository_locations.bzl]: Thanks!

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 6]
```diff
@@ -572,5 +574,26 @@ TEST_F(QuicPlatformTest, FailToPickUnsedPort) {
   EXPECT_DEATH_LOG_TO_STDERR(QuicPickUnusedPortOrDie(), "Failed to pick a port for test.");
 }
 
+TEST_F(QuicPlatformTest, TestEnvoyQuicBufferAllocator) {
+  bool deterministic_stats = Envoy::Stats::TestUtil::hasDeterministicMallocStats();
+  size_t start_mem;
+  if (deterministic_stats) {
+    start_mem = Envoy::Memory::Stats::totalCurrentlyAllocated();
```
Comment (jmarantz) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: Might as well just assign start_mem conditionally.

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 7]
```diff
@@ -572,5 +574,26 @@ TEST_F(QuicPlatformTest, FailToPickUnsedPort) {
...
+    std::cerr << "start mem " << start_mem << "\n";
```
Comment (jmarantz) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: dd

Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: done

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc (hunk 6 above)
Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: done

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 8]
```diff
@@ -572,5 +574,22 @@ TEST_F(QuicPlatformTest, FailToPickUnsedPort) {
   EXPECT_DEATH_LOG_TO_STDERR(QuicPickUnusedPortOrDie(), "Failed to pick a port for test.");
 }
 
+TEST_F(QuicPlatformTest, TestEnvoyQuicBufferAllocator) {
+  bool deterministic_stats = Envoy::Stats::TestUtil::hasDeterministicMallocStats();
+  const size_t start_mem =
+      deterministic_stats ? Envoy::Memory::Stats::totalCurrentlyAllocated() : 0;
```
Comment (jmarantz) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: actually you don't need to conditionalize this; totalCurrentlyAllocated() will return 0 if deterministic_stats is false.

Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: done

**Pull request merged**